Pal 873 audit service streaming documentation (#281)
gchq/Palisade-services: audit-service/README.md

# Audit Service

> The Audit service accepts incoming messages on the `error` and `success` Kafka topics, these messages contain all the details of the initial request and any other relevant information. 

Reviewer: I think it needs a rewording.  The "messages contain all the details of the initial request and any other relevant information" is vague and arguably wrong.  Suggest something in your own word along the lines of "The message contains information about the  request and in the case of an error, information about the error".

Author: This has been re-worded

> This information will be passed to any local audit services that have been implemented. This service does not have any output Kafka topics.

Reviewer: Could do with more information on the local audit services on the high level.  Further along there ought to be a technical description.  Something  as a developer will help me understand what do I have to do to use this in my application?

Author: Added information about the interface at the start of the ReadMe

# Audit Service

> The Audit service is an implementation of the [AuditService Interface](src/main/java/uk/gov/gchq/palisade/service/audit/service/AuditService.java) which audits any incoming messages.

Reviewer 2: This sentence kind of doesnt make a lot of sense... especially as the first sentence in the readme
Id suggest something like
```suggestion
The Audit Service can have multiple implementations, configured in the [application.yaml](src/main/resources/application.yaml), but simply, all implementations recieve, and log Audit Messages from the other Palisade services
```

Author: Updated

The service accepts incoming messages on the `error` and `success` Kafka topics. These messages contain information about the original request and, depending on the message type, will contain either of the following:

1. In the case of an error message, all the details of the error. 
1. In the case of a success message, the id value of the returned resource.

This information will be passed to any local audit services that have been implemented. This service does not have any output Kafka topics.

## High level architecture

<!--- 
See audit-service/doc/audit-service.drawio for the source of this diagram
--->
![Audit Service diagram](doc/audit-service.png)

> ## Flow of Control

Reviewer 2: Can you colour co-ordinate the flow diagram, kafka topics in that purple/pink colour please!

Author: Updated colours

*The token value come from the headers of the Kafka message that the service receives. This links the audit message to the original request that was made.

If an error has occurred at any stage during either the request or read phases then an AuditErrorMessage will be added to the `error` Kafka topic by the service that encountered the issue. 

> This type of message can be sent from any of the Palisade services (e.g. User Service or Policy Service).

Reviewer: Needs to be clarified.  The text of "during either the request or read phases" could be better described.  It suggests that the error messages are limited in the context of what they report.  The error messages can occur in any step of processing the request throughout Palisade Services.   
The word any in "can be sent from any of the Palisade Services" should be replaced with all.   Suggest either listing all of the services or none as listing two is misleading.
Also, e.g. User Service or Policy Service

Author: I've updated this part so that it reads differently.

> This message will then be read by the Audit Service and passed onto the local Audit Service implementation to allow the details of the error to be logged.

Reviewer: This section needs more detail to explain what it looks like how it works.  The reader will be a developer who wants to know how they can implement this.  If we have an example or a test it would be helpful to point it out to explain.  Same goes for the back-up for messages that could not be deserialised.  

In addition, the the service package is important and a suggestion would be to add a discussion the classes in the service package to the documentation

Author: I have added information about the interface at the start of the ReadMe now. This section has also been re-worded.

*The token value come from the headers of the Kafka message that the service receives. This links the audit message to the original request that was made.

If an error has occurred while Palisade is processing a client request then an AuditErrorMessage will be added to the `error` Kafka topic by the service that encountered the issue. 

> This type of message can be sent from all the Palisade services.

Reviewer 2: ```suggestion
This type of message can be sent from any of the Palisade services.
```

Author: This was changed based on previous comments but I do think this works better. Will change back.

directory. The directory is configured by adding the value to the [application.yaml](src/main/resources/application.yaml) configuration file under `audit.errorDirectory`.
This value can be changed depending on the profile that is used when the service is started.

The name of the file will contain the topic name the message was read from, either "Success" or "Error", and it will also include the timestamp of when the file was created.
This is done from within the [SerDesConfig](src/main/java/uk/gov/gchq/palisade/service/audit/stream/SerDesConfig.java) and if there are any issues creating or saving the file
then an error is logged in the service logs

## Kafka Interface

The application will not receive any `START` or `END` messages on either the `success` or `error` Kafka topics. The `success` topic will only consist of AuditSuccessMessage objects and the `error` topic will only consist of AuditErrorMessage objects. The
service will consume these messages and process them accordingly but there is no output from this service, instead it will acknowledge the incoming message so that it does not get processed more than once.

> The messages from each topic are processed by the [RunnableGraph](src/main/java/uk/gov/gchq/palisade/service/audit/stream/config/AkkaRunnableGraph.java). This takes the deserialised message and passes it into the 

Reviewer 2: ```suggestion
The messages from each topic are processed by the [RunnableGraph](src/main/java/uk/gov/gchq/palisade/service/audit/stream/config/AkkaRunnableGraph.java). This takes the deserialised message and passes it to the 
```

Author: Updated

> [Audit Service Proxy](src/main/java/uk/gov/gchq/palisade/service/audit/service/AuditServiceAsyncProxy.java) to be checked and then audited by all the service implementations.

Reviewer 2: to be checked? checked for what?

Author: The proxy checks have been added.

## Rest Interface

> The application exposes two REST endpoints for the purpose of debugging:

Reviewer 2: ```suggestion
The Audit Service exposes two REST endpoints for the purpose of debugging:
```

Author: Updated

>     - accepts either an `AuditSuccessMessage` or an `AuditErrorMessage` as the request body

Reviewer: The core of the work flow for this is the AkkaRunnableGraph and all of it could be best described by walking though the runner method.  I think that there needs to be a section that describes this.

Author: This part is not about the RunnableGraph as this is how a developer can debug the service. I have added something about the RunnableGraph within the Kafka Interface section above here.
